tests/unit/test_types.py

```diff
@@ -208,9 +212,405 @@ def test_async_api_spec_from_and_to_dict(faker: Faker):
                 },
             }
         },
-        "servers": {"development": {"url": "localhost", "protocol": "ws"}},
+        "servers": {
+            "development": {
+                "url": "localhost",
+                "protocol": "ws",
+                "security": [{"test": []}],
+            }
+        },
+        "components": {
+            "securitySchemes": {
+                "test": {"type": "http", "scheme": "basic"},
+                "test2": {"type": "http", "scheme": "bearer", "bearerFormat": "JWT"},
+                "testApiKey": {"type": "httpApiKey", "name": "test", "in": "header"},
+                "oauth2": {
+                    "type": "oauth2",
+                    "flows": {
+                        "implicit": {
+                            "authorizationUrl": "https://localhost:12345",
+                            "refreshUrl": "https://localhost:12345/refresh",
+                            "scopes": {"a": "A", "b": "B"},
+                        }
+                    },
+                },
+            }
+        },
     }
 
     spec = AsyncApiSpec.from_dict(data)
     assert isinstance(spec, AsyncApiSpec)
     assert spec.to_dict() == data
+
+
+def test_oauth2_implicit_flow_validation():
+    scopes = {"a": "A", "b": "B"}
+    # authorization_url is required for implicit flow
+    flow = OAuth2Flow(scopes=scopes, authorization_url=None)
+
+    with pytest.raises(ValueError):
+        OAuth2Flows(implicit=flow)
+
+
+def test_oauth2_password_flow_validation():
+    scopes = {"a": "A", "b": "B"}
+    # token_url is required for password flow
+    flow = OAuth2Flow(scopes=scopes, token_url=None)
+
+    with pytest.raises(ValueError):
+        OAuth2Flows(password=flow)
+
+
+def test_oauth2_client_credentials_flow_validation():
+    scopes = {"a": "A", "b": "B"}
+    # token_url is required for client_credentials flow
+    flow = OAuth2Flow(scopes=scopes, token_url=None)
+
+    with pytest.raises(ValueError):
+        OAuth2Flows(client_credentials=flow)
+
+
+def test_oauth2_authorization_code_flow_validation():
+    scopes = {"a": "A", "b": "B"}
+    # token_url is required for authorization_code flow
+    flow = OAuth2Flow(scopes=scopes, token_url=None)
+
+    with pytest.raises(ValueError):
+        OAuth2Flows(authorization_code=flow)
+
+
+def test_security_scheme_validation():
+    with pytest.raises(ValueError):
+        # missing flows
+        SecurityScheme(type=SecuritySchemesType.OAUTH2)
+
+    with pytest.raises(ValueError):
+        # missing flows
+        SecurityScheme(type=SecuritySchemesType.OPENID_CONNECT)
+
+    with pytest.raises(ValueError):
+        # missing scheme
+        SecurityScheme(type=SecuritySchemesType.HTTP)
+
+    with pytest.raises(ValueError):
+        # missing in
+        SecurityScheme(type=SecuritySchemesType.HTTP_API_KEY)
+    with pytest.raises(ValueError):
+        # invalid in
+        SecurityScheme(type=SecuritySchemesType.HTTP_API_KEY, in_="garbage")
```

Comment:
Not sure if this test is relevant anymore, since we're now using an enum rather than a string.